### PR TITLE
feat: enable .NET 7.0 support

### PR DESCRIPTION
**Pull request by vbreuss**
Update interfaces to support new functionality from .NET 7.
The implementation in the testing helper is not part of this PR!

### CODE REVIEW

File: src/TestableIO.System.IO.Abstractions/IPath.cs [hunk 1]
```diff
@@ -41,6 +41,11 @@ public interface IPath : IFileSystemEntity
 	    bool EndsInDirectorySeparator(string path);
 #endif
 
+#if FEATURE_FILESYSTEM_NET7
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IPath.cs:44]:
```suggestion
#if FEATURE_PATH_EXISTS
```

File: src/TestableIO.System.IO.Abstractions/IFileSystemWatcher.cs [hunk 2]
```diff
@@ -64,5 +64,10 @@ public interface IFileSystemWatcher : IFileSystemEntity, IDisposable
 
         /// <inheritdoc cref="FileSystemWatcher.WaitForChanged(WatcherChangeTypes, int)" />
         IWaitForChangedResult WaitForChanged(WatcherChangeTypes changeType, int timeout);
+
+#if FEATURE_FILESYSTEM_NET7
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IFileSystemWatcher.cs:68]:
```suggestion
#if FEATURE_FILE_SYSTEM_WATCHER_WAIT_WITH_TIMESPAN
```

File: src/TestableIO.System.IO.Abstractions/IDirectory.cs [hunk 3]
```diff
@@ -10,12 +10,22 @@ public interface IDirectory : IFileSystemEntity
...
+	/// <inheritdoc cref="Directory.CreateDirectory(string, UnixFileMode)" />
+	IDirectoryInfo CreateDirectory(string path, UnixFileMode unixCreateMode);
+#endif
+
 #if FEATURE_CREATE_SYMBOLIC_LINK
         /// <inheritdoc cref="Directory.CreateSymbolicLink(string, string)" />
         IFileSystemInfo CreateSymbolicLink(string path, string pathToTarget);
 #endif
 
+#if FEATURE_FILESYSTEM_NET7
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IDirectory.cs:24]:
```suggestion
#if FEATURE_CREATE_TEMP_SUBDIRECTORY
```

File: src/TestableIO.System.IO.Abstractions/IDirectory.cs [hunk 4]
```diff
@@ -10,12 +10,22 @@ public interface IDirectory : IFileSystemEntity
     {
         /// <inheritdoc cref="Directory.CreateDirectory(string)" />
         IDirectoryInfo CreateDirectory(string path);
-        
+
+#if FEATURE_FILESYSTEM_UNIXFILEMODE
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IDirectory.cs:14]:
```suggestion
#if FEATURE_UNIX_FILE_MODE
```

File: src/TestableIO.System.IO.Abstractions/IFile.Async.cs [hunk 5]
```diff
@@ -53,6 +53,19 @@ partial interface IFile
             Encoding encoding,
             CancellationToken cancellationToken = default);
 
+#if FEATURE_FILESYSTEM_NET7
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IFile.Async.cs:56]:
```suggestion
#if FEATURE_READ_LINES_ASYNC
```

File: src/TestableIO.System.IO.Abstractions/IFile.cs [hunk 6]
```diff
@@ -67,25 +68,71 @@ public interface IFile : IFileSystemEntity
 
         /// <inheritdoc cref="File.GetAttributes(string)" />
         FileAttributes GetAttributes(string path);
+#if FEATURE_FILESYSTEM_SAFEFILEHANDLE
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions/IFile.cs:71]:
```suggestion
#if FEATURE_FILE_ATTRIBUTES_VIA_HANDLE
```

File: Directory.Build.props [hunk 7]
```diff
@@ -11,9 +11,10 @@
...
+    <DefineConstants Condition="'$(TargetFramework)' == 'net7.0' OR '$(TargetFramework)' == 'net6.0' OR '$(TargetFramework)' == 'net5.0' OR '$(TargetFramework)' == 'netcoreapp3.1' OR '$(TargetFramework)' == 'netstandard2.1'">$(DefineConstants);FEATURE_ASYNC_FILE;FEATURE_ENUMERATION_OPTIONS;FEATURE_ADVANCED_PATH_OPERATIONS;FEATURE_PATH_JOIN_WITH_SPAN;FEATURE_SPAN</DefineConstants>
+    <DefineConstants Condition="'$(TargetFramework)' == 'net7.0' OR '$(TargetFramework)' == 'net6.0' OR '$(TargetFramework)' == 'net5.0'">$(DefineConstants);FEATURE_FILE_MOVE_WITH_OVERWRITE;FEATURE_SUPPORTED_OS_ATTRIBUTE;FEATURE_FILE_SYSTEM_WATCHER_FILTERS;FEATURE_ENDS_IN_DIRECTORY_SEPARATOR;FEATURE_PATH_JOIN_WITH_PARAMS;FEATURE_PATH_JOIN_WITH_FOUR_PATHS</DefineConstants>
+    <DefineConstants Condition="'$(TargetFramework)' == 'net7.0' OR '$(TargetFramework)' == 'net6.0'">$(DefineConstants);FEATURE_FILE_SYSTEM_INFO_LINK_TARGET;FEATURE_CREATE_SYMBOLIC_LINK;FEATURE_FILESTREAM_OPTIONS</DefineConstants>
+    <DefineConstants Condition="'$(TargetFramework)' == 'net7.0'">$(DefineConstants);FEATURE_FILESYSTEM_NET7;FEATURE_FILESYSTEM_SAFEFILEHANDLE;FEATURE_FILESYSTEM_UNIXFILEMODE</DefineConstants>
```
Comment (fgreinacher) [inline Directory.Build.props:17]: Let's try to make these more specific to follow the existing pattern. I commented my suggestions below on the interfaces. 

(I know it's a bit cumbersome, but consistency is everything 😀)

File: src/TestableIO.System.IO.Abstractions.TestingHelpers/MockDirectory.cs [hunk 8]
```diff
@@ -39,6 +39,14 @@ public override IDirectoryInfo CreateDirectory(string path)
             return CreateDirectoryInternal(path);
         }
 
+#if FEATURE_FILESYSTEM_UNIXFILEMODE
+        /// <inheritdoc />
+        public override IDirectoryInfo CreateDirectory(string path, UnixFileMode unixCreateMode)
+        {
+            throw new NotImplementedException();
```
Comment (fgreinacher) [inline src/TestableIO.System.IO.Abstractions.TestingHelpers/MockDirectory.cs:46]: Should we introduce some helper method for all these `NotImplementedException`s that tells people what is not supported and point them to the issue tracker for requesting it?

**Review by fgreinacher — CHANGES_REQUESTED** [review]
Thanks @vbreuss, good to have support for .NET 7 👍

This looks great, I just have some stylistic suggestions.

File: Directory.Build.props (hunk 7 above)
Comment (vbreuss) [reply Directory.Build.props:17]: No problem, I renamed the feature flags as to your suggestions.

File: src/TestableIO.System.IO.Abstractions.TestingHelpers/MockDirectory.cs (hunk 8 above)
Comment (vbreuss) [reply src/TestableIO.System.IO.Abstractions.TestingHelpers/MockDirectory.cs]: I replaced all empty `NotImplementedExceptions` in the testinghelpers project with
```csharp
new NotImplementedException(StringResources.Manager.GetString("NOT_IMPLEMENTED_EXCEPTION"));
```

Comment (fgreinacher) [reply src/TestableIO.System.IO.Abstractions.TestingHelpers/MockDirectory.cs]: Oh, that's nice, I totally forgot about that one 👍

**Review by fgreinacher — APPROVED** [review]
Thanks 🎉

**Pull request merged**